### PR TITLE
AVRO-4161: [C#] Fix System.Decimal to Avro decimal conversion

### PR DESCRIPTION
**Pull request by rayokota**
## What is the purpose of the change

The conversion from the C# System.Decimal to an Avro logical type decimal is not correct. The first 12 bytes of the System.Decimal represent a positive unscaled value. This value is passed into a BigInteger constructor. However, the documentation of the [BigInteger constructor](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.biginteger.-ctor?view=net-9.0#system-numerics-biginteger-ctor(system-byte())) states that the most significant bit of the last byte represents the sign of the BigInteger. The docs also explain that a 0 byte can be appended to the byte array passed to the BigInteger constructor to ensure the value is treated as positive.

## Verifying this change

Added a test that fails without the fix.

## Documentation

- Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by rayokota** [comment]
@KalleOlaviNiemitalo @zcsizmadia could I kindly get a review? Thanks in advance!

**Comment by KalleOlaviNiemitalo** [comment]
This looks OK but I have doubts about AvroDecimal.CompareTo

**Comment by KalleOlaviNiemitalo** [comment]
Filed the AvroDecimal.CompareTo bug as [AVRO-4162](https://issues.apache.org/jira/browse/AVRO-4162).

**Comment by rayokota** [comment]
> Filed the AvroDecimal.CompareTo bug as [AVRO-4162](https://issues.apache.org/jira/browse/AVRO-4162).

I can take a look at this after this PR is merged

**Comment by KalleOlaviNiemitalo** [comment]
If you're interested in fixing more AvroDecimal conversion bugs, there's also the infinite recursion [AVRO-3569](https://issues.apache.org/jira/browse/AVRO-3569).

File: lang/csharp/src/apache/test/AvroDecimalTest.cs [hunk 1]
```diff
@@ -38,7 +38,19 @@ public void TestAvroDecimalToString(decimal value)
...
+            var avroDecimal = new AvroDecimal(value);
+            var avroDecimalString = avroDecimal.ToString();
+
+            Assert.AreEqual(valueString, avroDecimalString);
+        }
+
+        [Test]
+        public void TestHighPrecisionAvroDecimalToString()
+        {
+            var value = 4.1748330066797328106875724512m; // High precision decimal value
```
Comment (martin-g) [inline lang/csharp/src/apache/test/AvroDecimalTest.cs:50]: What is `m` at the end of the number ? Magic ? :-)

Comment (KalleOlaviNiemitalo) [reply lang/csharp/src/apache/test/AvroDecimalTest.cs:50]: It's C# syntax for a constant of type `decimal`.

Comment (KalleOlaviNiemitalo) [reply lang/csharp/src/apache/test/AvroDecimalTest.cs:50]: Specified in [C# 5 section 7.4.5.4 Real literals](https://github.com/dotnet/csharpstandard/blob/V5-standard/standard/lexical-structure.md#7454-real-literals).

**Comment by rayokota** [comment]
@martin-g @zcsizmadia , anything else I need to do for this PR?  Thanks!

**Comment by martin-g** [comment]
Just wait ...
I guess @zcsizmadia will take a look soon.

File: lang/csharp/src/apache/test/AvroDecimalTest.cs (hunk 1 above)
Comment (KalleOlaviNiemitalo) [inline lang/csharp/src/apache/test/AvroDecimalTest.cs:50]: Can you add a test of converting a negative `decimal` to AvroDecimal?  Even though this PR doesn't change how that works.

**Review by zcsizmadia — APPROVED** [review]

File: lang/csharp/src/apache/test/AvroDecimalTest.cs (hunk 1 above)
Comment (rayokota) [reply lang/csharp/src/apache/test/AvroDecimalTest.cs:50]: Sure, I added a test for a negative high precision value

**Comment by rayokota** [comment]
@martin-g , I added a requested test, would it be possible for you to kick off another build?  Thanks!
